#559 Address Retry ordering issues and refactor the logic

anilgursel opened the pull request:
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [x] Tests added.
- [x] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

**Description:**

* Fixed ordering issue for retried elements.
* Fixed not completing the stage until there is further demand from downstream.
* A retried element with a lower retry time should be able to go to the head of the retry queue
* Optimizations on method calls and map accesses.
* Stream should complete when upstream completes. Should not require a new demand from downstream to complete (the tests were requiring an extra demand from sink).
* Dropped Overflow strategy support as anything other than backpressure  does not make sense for Retry.  It is not really a buffer, any element in retryregistry is actually inflight, which we cannot drop.


**Things to do:**

* Add test case for putting an element to the head of the queue when retry time is sooner than head's.
* Go over onDownstreamFinish scenarios.

anilgursel commented on squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala:
There is a potential of stream stuck here..  if onPush(in1) cannot push and element remains at the port, we probably need to grab the element here.

anilgursel commented on squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala:
I agree.  That would prevent unnecessary scheduling of the timer.

akara commented on squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala:
Function name does not communicate that well. Can we make it `shouldPullIn1`? So to use it this will look like

```scala
else if (shouldPullIn1) pull(in1)
```

akara commented on squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala:
```diff
@@ -401,38 +338,22 @@ final class RetryBidi[In, Out, Context] private[streams](maxRetries: Int, unique
+      if (isAvailable(out1)) {
+        val (elem, ctx, _) = retryRegistry(retryQ.dequeue()._2)
+        push(out1, (elem, ctx))
+      }
```
Normally, if there was a demand in `in2`, `out1` should automatically be available. Is there any chance a component on the `out1` -> `in2` loop would cause back-pressure making the demand not propagate? A common test case is a `ClientFlow` call which does not properly consume or discard its entity.

At this time we only rely on an `onPull` on `out1` after this point, and no longer a scheduler. Perhaps we should leave a comment.

anilgursel replied:
Since `Retry` does not control the joined flow, we cannot guarantee the propagation.  Also, the propagation is async.  I am not clear what comment we should leave?  You mean something like below?

```scala
final override def onTimer(key: Any): Unit = {
  if (isAvailable(out1)) {
    val (elem, ctx, _) = retryRegistry(retryQ.dequeue()._2)
    push(out1, (elem, ctx))
  } // else push the element when the demand arrives with the next onPull
```

anilgursel commented:
@akara 

> Cancelling the scheduler and just rely on the next onPull is a good strategy. But it needs proper documentation.

You mean documentation as in comments or in the actual user facing documentation.  Since it is implementation detail, I'd prefer it to be in the code.  

Please see if the latest commit addresses your comments.
